contextual_lenses/train_utils.py

```diff
@@ -0,0 +1,136 @@
+"""Train utils
+
+General tools for instantiating and training models."""
+
+
+import flax
+from flax import nn
+from flax import optim
+from flax.training import checkpoints
+from flax.training import common_utils
+
+import jax
+from jax import random
+import jax.nn
+import jax.numpy as jnp
+
+import numpy as np
+
+import functools
+
+
+def create_optimizer(model, learning_rate, weight_decay):
+  """Instantiates Adam optimizer."""
+
+  optimizer_def = optim.Adam(learning_rate=learning_rate, weight_decay=weight_decay)
+  optimizer = optimizer_def.create(model)
+  
+  return optimizer
+
+
+@functools.partial(jax.jit, static_argnums=(3, 4))
+def train_step(optimizer, X, Y, loss_fn, loss_fn_kwargs):
+  """Trains model (optimizer.target) using specified loss function."""
+
+  def compute_loss_fn(model, X, Y, loss_fn, loss_fn_kwargs):
+    Y_hat = model(X)
+    loss = loss_fn(Y, Y_hat, **loss_fn_kwargs)
+    return loss
+  
+  grad_fn = jax.value_and_grad(compute_loss_fn)
+  _, grad = grad_fn(optimizer.target, X, Y, loss_fn, loss_fn_kwargs)
+  optimizer = optimizer.apply_gradient(grad)
+  
+  return optimizer
+
+
+def train(model, train_data, loss_fn, loss_fn_kwargs, learning_rate=1e-4, weight_decay=0.1,
+          restore_dir=None, save_dir=None):
+  """Instantiates optimizer, applies train_step over training data.""" 
+  
+  optimizer = create_optimizer(model, learning_rate=learning_rate, weight_decay=weight_decay)
+
+  if restore_dir is not None:
+    optimizer = checkpoints.restore_checkpoint(ckpt_dir=restore_dir, target=optimizer)
+    
+  for batch in iter(train_data):
+    X, Y = batch
+    optimizer = train_step(optimizer, X, Y, loss_fn, loss_fn_kwargs)
+  
+  if save_dir is not None:
+    checkpoints.save_checkpoint(ckpt_dir=save_dir, target=optimizer, step=optimizer.state.step)
+
+  return optimizer
+
+
+p_train_step = jax.pmap(train_step, axis_name='batch', static_broadcasted_argnums=(3, 4))
+
+
+def p_train(model, train_data, loss_fn, loss_fn_kwargs, learning_rate=1e-4, weight_decay=0.1,
```

Comment:
lots of code duplication. Can you just have a flag called 'use_pmap' for train()?